adr/ADR-15.md

```diff
@@ -127,71 +127,114 @@ If the result indicates that the consumer already exists, then it means that the
+The implementation of ordered consumers will possibly want to handle detection of gaps itself, 
+so clients may want to provide for optional detection of message gaps.
+
+#### Flow Control
 
-The server may include, in heartbeats status messages, a header called `Nats-Consumer-Stalled`, with a value being a flow control subject. If this header is found in an hearbeat message, the library should send "right away" an empty message to the flow control subject, regardless of its current flow control state.
+When a JetStream consumer has flow control enabled (not applicable to pull consumers), the library may receive a flow control status message instead of a regular JetStream message.
+
+At the time when the message would have been normally passed to the user for processing, either synchronously via "Next Message" or via an asynchronous callback, 
+the client should respond to the provided reply-to by publishing the value as the subject for a message with an empty payload.
+
+When a flow control message is reached, the server likely had sent more messages.
+If the user is requesting messages synchronously via Next Message, the client should try to respond with the next buffered message after it responds to the flow control, 
+if it is within any wait time supplied by the user as part of the next message call.     
+
+**Note:** The format of the Flow Control subject should not to be inspected, since it may change at the server discretion, but it will always be a publishable subject.
+
+It is possible, that either the flow control or its response is missed and that the consumer is considered stalled from a server perspective.
+So the server requires that when flow control is turned on, an idle heartbeat is set.
+If a flow control message is missed, heartbeat messages will contain a header called `Nats-Consumer-Stalled`, 
+the value being identical to the flow control subject.
+When the client detects a heartbeat with the stalled header, it publish to the server as it would responding to flow control.
 
-Without this, it can be that a consumer remains stalled, so having both heartbeats and flow control configured seem like necessary, so we may make that a misconfiguration if flow control is configured without heartbeat.
+It's possible that because of the idle heartbeat duration, that both flow control and multiple heartbeat messages contain the
+flow control subject. It is only required to respond to the specific subject once, so it is suggested that clients track 
+the last flow control subject responded to avoid replying multiple times for the same flow control. The server will ignore duplicates, it is just unnecessary traffic.  
 
-Again, Ordered consumers should be favored.
+#### Pull Mode Statuses
 
-**Note**: In most libraries the flow control messages are never presented to the user. Those are handled internally.
+In pull mode, 404 and 408 statuses should not be surfaced to the user or considered in fetch or iterate counts.
```

Comment:
Any status in general, there is also 409... then also 503 becomes an error etc...

Reply:
@wallyqs What are 409 and 503 for?

Reply:
409 - doing conflicting things like trying to pull from a push consumer, exceeding max ack pending etc
503 - no responder, like if you try to reply to fc and the subscribe went away

Reply:
So it seems like 409 and 503 should not be passed, but an error should be raised (not passed) to the user, unlike 404 and 408 which can be discarded. Updating the doc. Should we assume that unknown statuses should be treated as errors?